compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/gen/ArithmeticLIRGeneratorTool.java

```diff
@@ -100,7 +100,9 @@
 
     Variable emitLoad(LIRKind kind, Value address, LIRFrameState state);
 
-    void emitStore(ValueKind<?> kind, Value address, Value input, LIRFrameState state);
+    Variable emitVolatileLoad(LIRKind kind, Value address, LIRFrameState state);
+
+    void emitStore(ValueKind<?> kind, Value address, Value input, LIRFrameState state, boolean isVolatile);
```

Comment:
I feel a bit uneasy by this API change. The `isVolatile` flag is only relevant for a single platform. But API suggests that this call always emits a volatile store, which is wrong. An assertion in the backends would help here.

I would go for adding an assertion to each platform (`LLVM`, `SPARC` and `AMD64`) that this flag must be false since they never emit volatile stores.

This also requires to replace the `WriteNode` in the `InsertMembarsPhase` with a new node having `volatileAccess==false`. Then we can add an assertion to all these platforms which assumes `isVolatile==false`.

Reply:
Thanks for reviewing this. Wouldn't you then also have InsertMembarsPhase turn all VolatileReadNode into a regular ReadNode?

Reply:
Thx. I'll test it.

Reply:
BootstrapWithSystemAssertions fails with this exception:

```
gate: 06 Nov 2019 20:04:17(+03:04) BEGIN: BootstrapWithSystemAssertionsEconomy
env MX_SUBPROCESS_COMMAND_FILE=/.tmp/mx_subprocess_command.55OjMd MX_PRIMARY_SUITE_PATH=/b/b/e/main/compiler MX__SUITEMODEL=sibling \
/compiler/mxbuild/linux-amd64/graaljdks/jdk1.8-cmp/bin/java -server -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI -d64 -Dgraal.PrintGraph=Network -Dgraal.CompilationWatchDogStartDelay=60.0D -esa -Dgraal.DetailedAsserts=true -Dgraal.CompilerConfiguration=economy -Dgraal.CompilationFailureAction=ExitVM -XX:+UseJVMCICompiler -Dgraal.DumpOnError=true -Dgraal.PrintGraphFile=true -XX:-TieredCompilation -XX:+BootstrapJVMCI -version
Bootstrapping JVMCI[Use -Dgraal.LogFile=<path> to redirect Graal log output to a file.]
[thread:11] scope: JVMCI CompilerThread6
    [thread:11] scope: JVMCI CompilerThread6.Compiling.GraalCompiler
    Context: StructuredGraph:130{HotSpotMethod<HotSpotJVMCIRuntime.runtime()>}
          [thread:11] scope: JVMCI CompilerThread6.Compiling.GraalCompiler.BackEnd.EmitLIR.LIRGeneration
          Exception raised in scope JVMCI CompilerThread6.Compiling.GraalCompiler.BackEnd.EmitLIR.LIRGeneration: org.graalvm.compiler.graph.GraalGraphError: org.graalvm.compiler.debug.GraalError: should not reach here
          	at node: 12|VolatileRead
          	at org.graalvm.compiler.graph.GraalGraphError.transformAndAddContext(GraalGraphError.java:136)
          	at org.graalvm.compiler.core.gen.NodeLIRBuilder.doBlock(NodeLIRBuilder.java:378)
          	at org.graalvm.compiler.core.LIRGenerationPhase.emitBlock(LIRGenerationPhase.java:85)
          	at org.graalvm.compiler.core.LIRGenerationPhase.run(LIRGenerationPhase.java:75)
          	at org.graalvm.compiler.core.LIRGenerationPhase.run(LIRGenerationPhase.java:46)
          	at org.graalvm.compiler.lir.phases.LIRPhase.apply(LIRPhase.java:117)
          	at org.graalvm.compiler.lir.phases.LIRPhase.apply(LIRPhase.java:109)
          	at org.graalvm.compiler.core.gen.LIRCompilerBackend.emitLIR0(LIRCompilerBackend.java:155)
          	at org.graalvm.compiler.core.gen.LIRCompilerBackend.emitLIR(LIRCompilerBackend.java:113)
          	at org.graalvm.compiler.core.gen.LIRCompilerBackend.emitBackEnd(LIRCompilerBackend.java:83)
          	at org.graalvm.compiler.core.target.Backend.emitBackEnd(Backend.java:279)
          	at org.graalvm.compiler.core.GraalCompiler.compile(GraalCompiler.java:145)
          	at org.graalvm.compiler.core.GraalCompiler.compileGraph(GraalCompiler.java:129)
          	at org.graalvm.compiler.hotspot.HotSpotGraalCompiler.compileHelper(HotSpotGraalCompiler.java:227)
          	at org.graalvm.compiler.hotspot.HotSpotGraalCompiler.compile(HotSpotGraalCompiler.java:245)
          	at org.graalvm.compiler.hotspot.CompilationTask$HotSpotCompilationWrapper.performCompilation(CompilationTask.java:170)
          	at org.graalvm.compiler.hotspot.CompilationTask$HotSpotCompilationWrapper.performCompilation(CompilationTask.java:84)
          	at org.graalvm.compiler.core.CompilationWrapper.run(CompilationWrapper.java:179)
          	at org.graalvm.compiler.hotspot.CompilationTask.runCompilation(CompilationTask.java:334)
          	at org.graalvm.compiler.hotspot.HotSpotGraalCompiler.compileMethod(HotSpotGraalCompiler.java:144)
          	at org.graalvm.compiler.hotspot.HotSpotGraalCompiler.compileMethod(HotSpotGraalCompiler.java:109)
          	at jdk.vm.ci.hotspot.HotSpotJVMCIRuntime.compileMethod(HotSpotJVMCIRuntime.java:667)
          Caused by: org.graalvm.compiler.debug.GraalError: should not reach here
          	at org.graalvm.compiler.debug.GraalError.shouldNotReachHere(GraalError.java:47)
          	at org.graalvm.compiler.core.amd64.AMD64ArithmeticLIRGenerator.emitVolatileLoad(AMD64ArithmeticLIRGenerator.java:1167)
          	at org.graalvm.compiler.nodes.memory.VolatileReadNode.generate(VolatileReadNode.java:54)
          	at org.graalvm.compiler.core.gen.NodeLIRBuilder.emitNode(NodeLIRBuilder.java:482)
          	at org.graalvm.compiler.core.gen.NodeLIRBuilder.doRoot(NodeLIRBuilder.java:473)
          	at org.graalvm.compiler.core.gen.NodeLIRBuilder.doBlock(NodeLIRBuilder.java:376)
          	... 20 more
```

Commandline: `mx gate -t BootstrapWithSystemAssertionsEconomy`

Reply:
The economy configuration must be broken then because it doesn't emit membars for volatile accesses at lowering and it doesn't run the InsertMembarsPhase. I see 3 ways to fix this:

- run the economy configuration with LateMembars to false. AFAIU, we would not be able to run code compiled with the economy configuration and the regular configuration in the same process on aarch64 because correctness requires that we uniformly use either membars or acquire/release instructions. (maybe @adinn can confirm that)

- Add InsertMembarsPhase to the economy configuration but that doesn't seem like a good fit for an economy configuration

- Have all platform support volatile accesses at the LIR level. So x86 would emit a regular access instruction surrounded by whatever membar instructions a required.

Reply:
> correctness requires that we uniformly use either membars or acquire/release instructions. (maybe @adinn can confirm that)

Well, yes, strictly we really ought to to uniformly use either membars or acquire/release instructions. That's because the barrier model for ordering and the ldar/stlr model operate slightly differently so straight mixing can lead to failures like the Dekker problem. However, they can be made compatible (at least with current hardware) but only if you live with a slightly raised cost for the barrier-based version. The cost is adding a membar (dmb ish) before the ldr that implements a volatile load. n.b. that's as well as the acquire barrier (dmb ishsd) that follows the load.

That's what we do in OpenJDK to allow C1 barrier-based translation of volatile accesses and C2 stlr/ldar-based translation of volatile accesses to interoperate. The extra barrier at load is not needed to keep threads executing C1 code consistent. However, it ensures that threads executing a load via C1-generated code see the right values when other threads are executing stores using C2-generated code.

There is also the question of the various Unsafe intrinsics to consider though. We had to make sure they also operated using the correct behaviours.

Reply:
What about the new patch that I just pushed? With LateMembars on, behavior of lowering depends on the platform. On aarch64, lowering never adds barriers (whether compiled under the economy or community configuration). All generated code uses the new barrier less volatile access scheme and there's no consistency issue between compiler configurations. On x86/sparc, lowering adds barriers for the economy configuration but not for the community configuration (barriers are then added late). This fixes the economy configuration that I broke with my previous change.

And, right, something needs to be done for Unsafe intrinsics. I'm not sure how to handle yet that so I'd like to keep it for a separate change.

Reply:
That doesn't play well with native image. I'm trying to find a cleaner way to do this.